### PR TITLE
Match Avatar Initial's color algorithm on MacOS with iOS/Android

### PR DESCRIPTION
**Pull request by kubalani**
### Platforms Impacted
- [ ] iOS
- [X] macOS

### Description of changes

Fix to issue #392 . Updating algorithms used to select foreground and background to match on other platforms. 

Screen Shot 2021-03-24 at 7 44 05 PM

### Verification

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Screen Shot 2021-03-24 at 5 39 33 PM | Screen Shot 2021-03-24 at 9 19 25 PM |

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
| Screen Shot 2021-03-24 at 9 23 12 PM| Screen Shot 2021-03-24 at 9 21 45 PM
 |

### Pull request checklist

This PR has considered:
- [X] Light and Dark appearances
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/491)

### CODE REVIEW

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (harrieshin) [inline macos/FluentUI/AvatarView/AvatarView.swift:450]: we should take one step further and actually have a .swift file that is shared between ios and mac projects

Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:450]: we have a lot of scope to share code across both repos, the handling of pod spec and framework is beyond the scope of this PR. This PR is required to update devmain code for MacOS Avatar view that I intend to merge by Monday to unblock other PR's

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (Saadnajmi) [inline macos/FluentUI/DynamicColor/DynamicColor.swift:34]: Completely off topic for this review, but if you wanted to make DynamicColor.swift be an actual NSColor instead of an NSObject, you can look into what we do in react-native-macos (https://github.com/microsoft/react-native-macos/blob/master/React/Base/macOS/RCTDynamicColor.m)
It's not the prettiest though :P

**Review by Saadnajmi — APPROVED** [review]

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (harrieshin) [inline macos/FluentUI/DynamicColor/DynamicColor.swift:31]: can you give me context why this is public now?

Comment (kubalani) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: This logic could be reused in devmain to detect theme changes instead of duplicating code in devmain.

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (kubalani) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:34]: Thanks Saad, I really like the idea. And code is not that bad :)

I initially had this as an extension but then I changed it to `NSObject` to have parity with iOS. At some point I would like to consolidate code across platforms for reusability and keep the interface same helps to accomplish it down the line.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (harrieshin) [inline macos/FluentUI/AvatarView/AvatarView.swift:435]: does this has to be var instead of let?

Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: its mutated on line 439 and 441 so this needs to be a var. If you are referring to initializing this one based on condition then yes we can. Since the condition to initialize this is complex I am concerned that would make code readability much harder. Is there any other specific reason you would to make it immutable?

Comment (harrieshin) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: the recommendation is always let unless there is a reason to be var in swift. Isn't combined only set once depending on which if statement ?

Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: if `combined` is defined as a `let` variable then it won't be accessible outside of its enclosing if/else. That's the only reason for using `var`

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (Saadnajmi) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: I think we have similar code in devmain that already does this. Unless there was a specific scenario that needed this, I would not make it public unless necessary.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (Saadnajmi) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]:
```swift
let combined: String = {
  //... logic
}()
```

Comment (Saadnajmi) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: ^ I think that works. I feel like I've used it at least once

Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: Sync'ed with Saad offline the above conditional initialization didn't work. @Harrie you have any other ideas?

Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: @kubalani if you just change it to let it'll work because you're initializing it in one of the 3 paths of the if statement. And it's not changed afterwards.

```
                let combined: String
		if let secondaryText = secondaryText, let primaryText = primaryText, secondaryText.count > 0 {
			combined = primaryText + secondaryText
		} else if let primaryText = primaryText {
			combined = primaryText
		} else {
			combined = ""
		}
```

Alternatively you can also do the closure initialization (as @Saadnajmi suggested):
Not sure what didn't work there?

```
		let combined: String = {
			if let primaryText = primaryText {
				if let secondaryText = secondaryText, secondaryText.count > 0 {
					return primaryText + secondaryText
				}

				return primaryText
			}

			return ""
		}()
```

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:450]: Although I understand that refactoring this code and start sharing more across mac/ios is beyond the scope of this task, we really need to start laying out foundation for that.
Do we have a task in ADO to track that?

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [inline macos/FluentUI/AvatarView/AvatarView.swift:344]: This is being put back for the Outlook scenario only? Or all devmain apps will still use it?

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [inline macos/FluentUI/AvatarView/AvatarView.swift:328]: Does this one still need to be public? It seems like it's not used in devmain.

Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:328]: Line 360 has a deprecation directive pointing to this one which is also deprecated.

`	@available(*, deprecated, message: "Use getColorForIndex instead")`

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [inline macos/FluentUI/AvatarView/AvatarView.swift:334]: Is this a table index or just a hash?
The Int returned by this function would have to divided by the table entry count and the remainder (mod) of it would be the index of the table, correct?

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [inline macos/FluentUI/AvatarView/AvatarView.swift:337]: Is it unique? If it's a hash being generated from the string, uniqueness is not guaranteed.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:435]: just replacing `let` with `var` did worked. Swift doesn't like the second initialization

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:344]: outlook scenario only

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:328]: thanks for catching this, updated the comment

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:334]: yes, I didn't alter the original comment but I think the author's intent was that.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:337]: I don't know for sure, I haven't modified the implementation of original method.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:344]: Based on your code review open in devmain, it seems like all the current uses in devmain will still use this, no?

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:334]: Let's make the comment accurate with what the code is doing!

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:337]: Let's remove "unique" from this comment. It may be misleading.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift]: Swift does not allow a second assignment because "let" define a constant that can only be initialized (first assignment).
You can still declare the constant and initialize it later.
Swift will also catch if you try to use a constant before it's initialized.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:450]: Is there an ADO item to track that?

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (rdeassis) [reply macos/FluentUI/AvatarView/AvatarView.swift:328]: The documentation comment seems a bit off too:
"Get the **color** associated with a given index"

it's actually returning a ColorSet object. Can you improve the comments there?

Also, what's the internal visibility comment about?
"- note: Internal visibility exists only for unit testing"

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (rdeassis) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: Yes! Let's not make it public or @objc visible unless it's needed!
This will always have an impact on the binary size.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:344]: yes, after we merge this PR these methods will still be required for outlook but rest of cocoaui will use the new API's

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (kubalani) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: I've reached out to Saad offline to see if there is any alternative. This method is required in devmain to retrieve theme variant of NSColor color.

Comment (kubalani) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: @Saadnajmi  do you have any objections to exposure to these methods? This method is required for because cocoaui consumes color API for their avatar view rendering

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift]: I have updated the description of the method, the mod usage is already described in note section of the comment.

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift]: I have removed unique from the description.

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (Saadnajmi) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: If we have to, then we have to 🤷🏾
I would file a bug on cocoaui to make sure they request the color on light/dark changes, even if we provide the right colors for them with this API regardless

File: macos/FluentUI/AvatarView/AvatarView.swift
Comment (kubalani) [reply macos/FluentUI/AvatarView/AvatarView.swift:326]: I have updated the comment. The note is no longer valid because I changed the test case in previous iteration.

Screen Shot 2021-04-01 at 8 44 26 AM
 
https://github.com/microsoft/fluentui-apple/pull/455/files#diff-48e9b1830c83b38133ae54b9dca476f4ab456571c9d162135bd0ac6e1ff65c8f

File: macos/FluentUI/DynamicColor/DynamicColor.swift
Comment (kubalani) [reply macos/FluentUI/DynamicColor/DynamicColor.swift:31]: I discussed offline with @Saadnajmi and @rdeassis  offline the possibility of using asset catalog to resolve colors. 
https://developer.apple.com/documentation/xcode/supporting_dark_mode_in_your_interface?language=objc

This would increase binary size as we need to duplicate colors. I am resolving this conversation and I have created another ADO task to continue this discussion. Please feel free to reopen if there are other concerns with the use of `DynamicColor`

**Review by rdeassis — APPROVED** [review]

**Pull request merged**
